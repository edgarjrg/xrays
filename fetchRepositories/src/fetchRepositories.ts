import { lensProp, lensPath, view, map, compose, flip, append, transduce } from 'ramda';
import { DoWhileAsync } from "../../repositories/src/DoWhileAsync";
import { BitbucketAPI } from "../../repositories/src/BitbucketAPIs";
import { Auth } from "../../auth/src";

export async function fetchAllPRLinksFrom(callback, repository: IRepository) {
  const bitbucketAPI = new BitbucketAPI(new Auth())

  let page = 1

  const response = await DoWhileAsync(
    () => bitbucketAPI.pullRequests(repository, 'MERGED', page),
    (httpResponse) => httpResponse.next !== undefined,
    (httpResponse) => {
      return callback(httpResponse2ObjectToPreserve(repository, httpResponse))
        .then(() => {
          page++;
        })
    }
  )
}

export function httpResponse2ObjectToPreserve(repository, httpResponse) {
  const links2ObjectToPreserve = map((pullRequestURL) => ({
    ...repository,
    pullRequestURL
  }))

  const httpResponse2Links = map(view(linkLens))

  const transducer = compose(httpResponse2Links, links2ObjectToPreserve)

  return transduce(transducer, flip(append), [], view(valuesLens, httpResponse))
}

export const valuesLens = lensProp('values')
export const linkLens = lensPath<any, string>(['links', 'self', 'href'])