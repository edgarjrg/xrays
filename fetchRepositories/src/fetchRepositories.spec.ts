import { valuesLens, linkLens, getLinks, fetchAllPRLinksFrom } from './fetchRepositories'
import { view } from 'ramda'

test('valuesLens should bring the value of response', async () => {
  const PRsResponse = require('./testData/mergedPRsResponse.json')
  const valuesLensResponse = require('./testData/valuesLensResponse.json')
  expect(view(valuesLens, PRsResponse)).toEqual(valuesLensResponse)
});

test('linkLens should bring the link of the PR', async () => {
  const linkLensInput = require('./testData/linkLens.input.json')
  const linkLensExpected = require('./testData/linkLens.expected.json')
  expect(view<string, any>(linkLens, linkLensInput)).toEqual(linkLensExpected)
});

test('getLinks should bring the link of the PR', async () => {
  const PRsResponse = require('./testData/mergedPRsResponse.json')
  expect(getLinks(PRsResponse)).toMatchSnapshot()
});



describe('fetchAllPRLinksFrom', () => {
  it('should brind all repos from a repository', async () => {
    const repo = {
      projectKey: "atlassian-oauth",
      repoSlug: "atlassian-oauth",
      userName: "atlassian",
    }
    const response = await fetchAllPRLinksFrom(repo)

    expect(response).toMatchSnapshot()
  });
});

