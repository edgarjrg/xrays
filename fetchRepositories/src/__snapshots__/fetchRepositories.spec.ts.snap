// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAllPRLinksFrom should bring all repository from a repository @integration @slow 1`] = `
Array [
  Array [
    Array [
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/35",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/34",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/33",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/32",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/31",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/30",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/29",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/28",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/26",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/25",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/24",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/23",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/21",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/19",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/18",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/17",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/15",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/16",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/14",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/13",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/11",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/10",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/9",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/3",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/8",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/7",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/6",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/5",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/4",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
      Object {
        "projectKey": "atlassian-oauth",
        "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/2",
        "repoSlug": "atlassian-oauth",
        "userName": "atlassian",
      },
    ],
  ],
]
`;

exports[`httpResponse2ObjectToPreserve 1`] = `
Array [
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/35",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/34",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/33",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/32",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/31",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/30",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/29",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/28",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/26",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
  Object {
    "projectKey": "atlassian-oauth",
    "pullRequestURL": "https://api.bitbucket.org/2.0/repositories/atlassian/atlassian-oauth/pullrequests/25",
    "repoSlug": "atlassian-oauth",
    "userName": "atlassian",
  },
]
`;
